lib/middleware/index.js

```diff
@@ -0,0 +1,106 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+/* jshint maxparams: 5 */
+
+var _ = require('underscore');
+
+/**
+ * A middleware is an object that implements any of the following methods:
+ *
+ * request(req, arg2, arg3, callback)
+ * response(req, err, res1, res2, callback)
+ *
+ * A middleware is free to alter req and update/replace any of arg2, arg3, err,
+ * res1 or res2. Multiple middlewares can be chained together forming a
+ * middleware stack. The callback is used to pass the control to the next
+ * middleware in chain. For example, a noop middleware can look like this:
+ *
+ * var noop = {
+ *     'request': function(req, arg2, arg3, callback) {
+ *         callback(arg2, arg3);
+ *     },
+ *     'response': function(req, err, res1, res2, callback) {
+ *         callback(err, res1, res2);
+ *     }
+ * };
+ *
+ * Given the middleware stack defined as:
+ *
+ * MiddlewareStack([m1, m2]).run(req, arg2, arg3, handler, handler_callback)
+ *
+ * The order in which the middlewares are applied in is: m1.request->
+ * m2.request-> handler-> m2.response-> m1.response-> handler_callback.
+ *
+ * The difference between applying a middleware stack to the server and the
+ * client is the logical direction a request and a response point to. For
+ * example, for a server it looks like this: server<-request, server->response;
+ * while for a client it's the reverse: client->request, client<-response.
+ */
+
+function MiddlewareStack(middlewares) {
+    this.middlewares = middlewares || [];
+}
+
+// XXX: adding nextTick to avoid blowing the call stack breaks a test
+MiddlewareStack.prototype.run = function run(req, arg2, arg3, handler, callback) {
+    var self = this;
+
+    var i = -1;
+    callRequestMiddleware(arg2, arg3);
+
+    function callRequestMiddleware(arg2, arg3) {
+        i += 1;
+        if (i < self.middlewares.length) {
+            var next = self.middlewares[i].request;
+            if (typeof next === 'function') {
+                next(req, arg2, arg3, callRequestMiddleware);
+            } else {
+                // skip this middleware if it doesn't implement request
+                callRequestMiddleware(arg2, arg3);
+            }
+        } else {
+            handler(req, arg2, arg3, callResponseMiddleware);
+        }
+    }
+    function callResponseMiddleware(err, res1, res2) {
+        i -= 1;
+        if (i >= 0) {
+            var next = self.middlewares[i].response;
+            if (typeof next === 'function') {
+                next(req, err, res1, res2, callResponseMiddleware);
+            } else {
+                // skip this middleware if it doesn't implement response
+                callResponseMiddleware(err, res1, res2);
+            }
+        } else {
+            callback(req, err, res1, res2);
+        }
+    }
+};
+
+
+module.exports = _.extend({
+        'MiddlewareStack': MiddlewareStack
+    },
+    require('./tombstone.js'),
```

Comment:
Do we need to include the actual middleware implementations here? Can we make this a "pure drop-in library"?

Maybe `index.js` could import all three modules, and this file (w/o imports) could become `core.js`?